amplitude.js

```diff
@@ -549,27 +613,71 @@ Amplitude.prototype.logRevenue = function(price, quantity, product) {
  * Remove events in storage with event ids up to and including maxEventId. Does
  * a true filter in case events get out of order or old events are removed.
  */
-Amplitude.prototype.removeEvents = function (maxEventId) {
-  var filteredEvents = [];
-  for (var i = 0; i < this._unsentEvents.length; i++) {
-    if (this._unsentEvents[i].event_id > maxEventId) {
-      filteredEvents.push(this._unsentEvents[i]);
+Amplitude.prototype.removeEvents = function (maxEventId, maxIdentifyId) {
+  if (maxEventId) {
+    var filteredEvents = [];
+    for (var i = 0; i < this._unsentEvents.length; i++) {
+      if (this._unsentEvents[i].event_id > maxEventId) {
+        filteredEvents.push(this._unsentEvents[i]);
+      }
+    }
+    this._unsentEvents = filteredEvents;
+  }
+
+  if (maxIdentifyId) {
+    var filteredIdentifys = [];
+    for (var j = 0; j < this._unsentIdentifys.length; j++) {
+      if (this._unsentIdentifys[j].event_id > maxIdentifyId) {
+        filteredIdentifys.push(this._unsentIdentifys[j]);
+      }
     }
+    this._unsentIdentifys = filteredIdentifys;
   }
-  this._unsentEvents = filteredEvents;
 };
 
 Amplitude.prototype.sendEvents = function(callback) {
-  if (!this._sending && !this.options.optOut && this._unsentEvents.length > 0) {
+  if (!this._sending && !this.options.optOut && this._unsentCount() > 0) {
     this._sending = true;
     var url = ('https:' === window.location.protocol ? 'https' : 'http') + '://' +
         this.options.apiEndpoint + '/';
 
     // Determine how many events to send and track the maximum event id sent in this batch.
-    var numEvents = Math.min(this._unsentEvents.length, this.options.uploadBatchSize);
-    var maxEventId = this._unsentEvents[numEvents - 1].event_id;
+    var numEvents = Math.min(this._unsentCount(), this.options.uploadBatchSize);
+
+    // coalesce events from both queues
+    var eventsToSend = [];
+    var eventIndex = 0;
+    var identifyIndex = 0;
+
+    while (eventsToSend.length < numEvents) {
+      var event;
+
+      // case 1: no identifys - grab from events
+      if (identifyIndex >= this._unsentIdentifys.length) {
+        event = this._unsentEvents[eventIndex++];
+
+      // case 2: no events - grab from identifys
+      } else if (eventIndex >= this._unsentEvents.length) {
+        event = this._unsentIdentifys[identifyIndex++];
+
+      // case 3: need to compare timestamps
+      } else {
+        if (this._unsentIdentifys[identifyIndex].timestamp <= this._unsentEvents[eventIndex].timestamp) {
```

Comment:
Yes it would be safer to have them share the same event queue, but I think that might confuse customers. They might see the event # sequence increasing, and then suddenly jump by 3 or 4. And the user properties don't show up in the stream, so they would think that 3 or 4 events got dropped or something. I think using timestamp is still pretty safe though.